### ISSUE TITLE
Duplicated modules in output file

### ISSUE THREAD

**Issue by avchugaev** [parcel-bundler/parcel#1838]
# 🐛 bug report

In my TypeScript + React [project](https://github.com/monumentjs/core/tree/master/examples/react-portlet) I've faced the problem with duplication of module in output JavaScript bundle. Same class imported two times.
Project is monorepo maintained with [Lerna](https://github.com/lerna/lerna).

## 🎛 Configuration (.babelrc, package.json, cli command)

**.babelrc**

```js
{
  "presets": [
    [
      "env",
      {
        "targets": {
          "browsers": [
            "last 1 Chrome version"
          ]
        }
      }
    ]
  ]
}
```

**package.json**

```json
{
  "private": true,
  "version": "0.9.0",
  "description": "React Portlet",
  "keywords": [],
  "scripts": {
    "serve": "parcel main/html/index.html -d build/debug",
    "build": "parcel build main/html/index.html -d build/release",
    "test": "td"
  },
  "dependencies": {
    "react": "^16.4.1",
    "react-dom": "^16.4.1",
    "@monument/core": "^0.9.0",
    "@monument/react": "^0.9.0"
  },
  "devDependencies": {
    "@types/react": "^16.4.6",
    "@types/react-dom": "^16.0.6",
    "parcel-bundler": "^1.9.7"
  }
}
```

**CLI command:**

```bash
$ npm run serve
```

## 🤔 Expected Behavior

I think that all modules should accur only once in output bundle.

## 😯 Current Behavior

Currenly some classes are duplicated in output bundle. 

## 💁 Possible Solution

I think root of the problem is symlink structure created by Lerna.

## 🔦 Context

In project I don't use code splitting yet.
At some point of time with help of logging I've found that some classes are duplicated in output bundle.

## 💻 Code Sample

In my case module with class "Class" is duplicated as well as probably some others.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.9.7
| Node             | 10.5.0
| npm         | 6.1.0
| Lerna | 2.11.0
| Operating System | Windows 10, Ubuntu

**Comment by achugaev93** [comment]
Possible problem is symlinks structure created by Lerna. Output javascript bundle includes same classes but under different file paths. "Unwrapping" symlinks may help in this situation. I mean paths should be pre-processed before registering a chunk to use real file path as chunk ID.

**Comment by achugaev93** [comment]
@DeMoorJasper , this bug is really blocker. Will you try to fix it in nearest future?

**Comment by jeanfortheweb** [comment]
Same Problem for me, module gehts executed twice.

**Comment by Offirmo** [comment]
Same problem for me, a monorepo handled by bolt, my modules are all duplicated, even when they have correct matching dependencies and peerPependencies.

**Comment by bertfreudenberg** [comment]
Same problem in our (non-react) monorepo. To illustrate, let's say we have 2 modules and 1 app:

```
modA:
   // no imports

modB:
  import "@comp/modA"

app: 
  import "@comp/modA"
  import "@comp/modB"
```

The folder structure created by `npm i` looks like this :
```
root
  modA
  modB
    node_modules
      @comp
        modA -> ../../../modA
  app
    node_modules
      @comp
        modA -> ../../../modA
        modB -> ../../../modB
```
But Parcel appears to not resolve the symlinks, it sees this as:
```
root
  modA
  modB
    node_modules
      @comp
        modA
  app
    node_modules
      @comp
        modA
        modB
          node_modules
            @comp
              modA

```
leading to `modA` being duplicated in the bundle:
```
"@comp/modA": ...
"@comp/modB": ...
"@comp/modB/node_modules/@comp/modA": ...
"app": ...
```
... which is obviously now what should happen.

**Comment by bertfreudenberg** [comment]
Btw, here's the workaround I'm using currently. I made a package (pasted below) and import that in every package that really needs to be deduplicated:

```
// Work around Parcel package duplication bug
// https://github.com/parcel-bundler/parcel/issues/1838

// Usage: import this in every file that needs deduplication
// Author: Bert Freudenberg (https://github.com/bertfreudenberg)

const parcel = module.bundle;

// deduplicate this, and every module that directly imports this
deduplicateImports([module.id, ...Object.keys(parcel.modules).filter(m => Object.values(parcel.modules[m][1]).includes(module.id))]);

function deduplicateImports(mods) {
    const modSources = mods.map(m => [m, ""+parcel.modules[m][0]]);
    const dupes = new Map();
    // find duplicates of given modules by comparing source code
    for (const dupe of Object.keys(parcel.modules)) {
        const dupeSource = ""+parcel.modules[dupe][0];
        for (const [mod, modSource] of modSources) {
            if (dupeSource === modSource && dupe !== mod) dupes.set(dupe, mod);
        }
    }
    // replace references to dupes with the actual modules
    const later = new Map();
    const fixed = new Set();
    for (const m of Object.values(parcel.modules)) {
        for (const [n, dupe] of Object.entries(m[1])) {
            const mod = dupes.get(dupe);
            if (mod && parcel.modules[mod]) {
                if (parcel.cache[dupe]) later.set(mod, dupe);   // dupe already loaded
                else {
                    m[1][n] = mod;                              // use mod
                    delete parcel.modules[dupe];                // delete dupe
                    fixed.add(n);
                }
            }
        }
    }
    for (const m of Object.values(parcel.modules)) {
        for (const [n, mod] of Object.entries(m[1])) {
            const dupe = later.get(mod);
            if (dupe && parcel.modules[dupe]) {
                m[1][n] = dupe;                                  // use dupe
                delete parcel.modules[mod];                      // delete mod
                fixed.add(n);
            }
        }
    }
    for (const fix of [...fixed].sort()) console.log(`Deduplicated import "${fix}"`);
}
```

**Comment by thecodermarcelo** [comment]
I'm using code splitting with react.lazy, and in routes as well, I made a workaround for this in my project which would be more easier in my case, I imported the icons that I had problems and put them together in a global file


```
//utils.js
import TopArrow from "Icons/general/whiteTopArrow.svg"
import DownArrow from "Icons/general/whiteDownArrow.svg"

window.icons = {
    ...window.icons,
    TopArrow,
    DownArrow
}
```
```
//index.js
import "utils.js"
```
```
//component.js
export default () => <img src={window.icons.TopArrow}/>
```
Hope it helps

**Comment by dharmax** [comment]
I'm experiencing the same issue. It appears out of nowhere, after i did some unrelated changes. 
It's with a js file. my fe router. Naturally, it kills my app and i still didn't find a working workaround.

**Comment by lrgalego** [comment]
> Btw, here's the workaround I'm using currently. I made a package (pasted below) which deduplicates everything that has the same source code:
> 
> ```
> // Work around Parcel package duplication bug
> // https://github.com/parcel-bundler/parcel/issues/1838
> 
> // Usage: import this to deduplicate modules
> // Author: Bert Freudenberg (https://github.com/bertfreudenberg)
> 
> // find duplicates of modules by comparing source code
> const parcel = module.bundle;
> const sources = {};
> const duplicates = {};
> for (const dupe of Object.keys(parcel.modules)) {
>     const source = "" + parcel.modules[dupe][0];
>     const id = sources[source];
>     if (id) duplicates[dupe] = id;
>     else sources[source] = dupe;
> }
> // replace references to duplicates with the actual modules
> const later = {};
> const fixed = new Set();
> for (const m of Object.values(parcel.modules)) {
>     for (const [n, dupe] of Object.entries(m[1])) {
>         const id = duplicates[dupe];
>         if (id) {
>             if (parcel.cache[dupe]) later[id] = dupe;      // dupe already loaded
>             else {
>                 m[1][n] = id;                              // use id
>                 delete parcel.modules[dupe];               // delete dupe
>                 fixed.add(n);
>             }
>         }
>     }
> }
> for (const m of Object.values(parcel.modules)) {
>     for (const [n, id] of Object.entries(m[1])) {
>         const dupe = later[id];
>         if (dupe) {
>             m[1][n] = dupe;                                 // use dupe
>             delete parcel.modules[id];                      // delete id
>             fixed.add(n);
>         }
>     }
> }
> for (const fix of [...fixed].sort()) console.log("Deduplicated import of", fix);
> ```

How do you use this package on your parcel?

**Comment by bertfreudenberg** [comment]
> @bertfreudenberg How do you use this package on your parcel?

Oh, you simply include it in your app as one of the first things to be imported/executed.

It's not a real fix, just a hacky work-around that works for us: parcel will still create a bundle with duplicate modules, but then when running the app my code goes through all modules replacing the duplicates. For this to work, my code needs to be executed before the duplicated modules (or at least before importing the duplicate).

**Comment by lrgalego** [comment]
That makes sense. I'm looking into a way to update the bundler before it packs, to apply the same logic you did on this package, but I haven't found a way (without forking parcel) to do this.
Let me know if you have any ideas on how to do it

**Comment by skipjack** [comment]
Seeing the same issue when `npm link`-ing a package where `react` is also installed. With `webpack`, you can customize resolution via the following configuration to fix this issue:

``` js
resolve: {
    modules: [ path.resolve(__dirname, 'node_modules'), 'node_modules' ]
}
```

Which essentially instructs `webpack` to always resolve packages in the main project's `node_modules` directory prior to looking in any other `node_modules` directory (e.g. the one in the linked package). I looked [here][1] but wasn't able to find any analogous configuration option for `parcel`.

This is really the only major blocker I've run into with `parcel`, otherwise it has been great and made projects much easier to get off the ground. IMO the behavior I described above is a pretty sensible default that would address the issue for most users (in the spirit of sensible defaults and little config). I'd be happy to submit a PR for this if the maintainers agreed.

[1]: https://parceljs.org/module_resolution.html

**Comment by skipjack** [comment]
It seems there is a workaround for this with [alias][1]:

__package.json__

``` json
  "alias": {
    "react": "./node_modules/react"
  }
```

The downside of this is that (I think) you would need to explicitly enter a line for each duplicated package. Webpack has a similar `alias` option that can be used the same way though the `modules` approach is a bit cleaner.

[1]: https://parceljs.org/module_resolution.html#aliases

**Comment by lrgalego** [comment]
@skipjack I created a repo to test the alias workaround you mentioned. but I got an interesting result.
When I run the dev server, the dependencies are duplicated in the bundle. (adding the alias fix the issue).
But when I run `parcel build` the dependencies are not added to the bundle. With our without the alias.
Here is my sample repo: https://github.com/lrgalego/parcel-test
That makes me believe that the code that trims the duplicated package is already in place, but it's not being invoked for the dev server.
I tested this with parcel 1.12.4

**Comment by jeanfortheweb** [comment]
i don't understand. i fixed this one years ago, including a pr with additional test cases. why is this still an issue?

**Comment by lrgalego** [comment]
@jeanfortheweb I sent a repo with a simple app showing the issue. maybe there's some edge case not covered by the test cases. 
https://github.com/lrgalego/parcel-test

**Comment by lrgalego** [comment]
@bertfreudenberg @skipjack 
Ok, I tried something different and it works like a charm.
Instead of making all the packages being linked to each other, I used yarn workspaces: https://classic.yarnpkg.com/en/docs/workspaces/

this way all the common dependencies are in the same directory and yarn symlink everything
so pkg1 looks for the dep on ../node_modules and pk2 also looks there. the path is the same and the dep is added only once in the bundle.
I created a new branch on my repo with the version that works well: https://github.com/lrgalego/parcel-test/tree/works-with-yarn-workspaces

**Comment by Offirmo** [comment]
Could that be added in parcel 2 roadmap?

**Comment by Offirmo** [comment]
@lrgalego are you sure your setup is complex enough?

The problem arise due to this:

* app
* node_modules/modA
* node_modules/modB
  * node_modules/modB/node_modules/modA (symlinked to modA above)

Due to require/import finding the closest matching module and then parcel seeing different pathes, modA could end up being duplicated even though it's the same module. Yarn workspaces doesn't help there.

**Comment by lrgalego** [comment]
What I have are all packages I wrote in a monorepo referencing each other by links. That's why .moving to yarn workspace, where the path is the same for all my packages solved the issue

**Comment by Offirmo** [comment]
@lrgalego I'm using workspaces as well, that's not the point. The modules are linked but since they have their own underlying node_modules, alternate pathes get picked. I edited my example above to make it clearer.

**Comment by danieltroger** [comment]
Hey guys, I'm also having the problem that parcel is including a file twice in the outputted bundle. I tried creating a module with just the file and adding it as dependency in the package.json file in the hopes that it only gets included once, but it's still twice in the output file.

And it sucks, because it adds 140KB to my final bundle :(

**Comment by danieltroger** [comment]
Just stumbled upon this duplication issue again, as seen in `--detailed-report`:

```
├── node_modules/lib/library.ts                                     7.72 KB    704ms
├── ../anothermoduleimade/node_modules/lib/library.ts                    7.72 KB    231ms

```

IMO this really should be fixed somehow that makes it impossible that the same file gets included twice, i.e. by comparing the filesize and if it matches calculating the hash of the contents and if the hash matches only including the file once.

**Comment by mischnic** [comment]
> by comparing the filesize of every file with the file size of every file and if one matches calculating the hash of the contents and if the hash matches only including the file once.

This would break some react libraries which have multiple files containing `import React from "react"; export default React.createContext();` and rely on them being not identical (`===`).

Does this still happen with Parcel 2?

**Comment by danieltroger** [comment]
@mischnic thanks for the reply. I don't know how to update to v2, is there a guide?

> This would break some react libraries which have multiple files containing import React from "react"; export default React.createContext(); and rely on them being not identical (===). #3523

That's actually funny because I rely on the behaviour of the modules being identical when imported from multiple places. So when I export an object I can modify it from one place I imported the module and it changes for all. Maybe it could be made configurable?

AFAIK if you need to clone an object you can return it from a function and call the function every time you want a copy.

So maybe my suggested deduplication paired with a configurable wrapping inside a function would work?

It could either be an option to the import statement but that's probably illegal or inside some config file or CLI option.

So you deduplicate everything for filesize reasons and then if people want to actually include two copies of the same file you just give them new copies in JS instead of sending 2.

**Comment by mischnic** [comment]
Rule of thumb: try it with Node's CJS behaviour or Node/browsers native ESM. This is how Parcel should behave, we are not trying to invent our own module format

**Comment by mischnic** [comment]
> is there a guide?

Using `parcel@nightly` instead of `parcel-bundler`.
https://v2.parceljs.org/getting-started/migration/

> new copies in JS instead of sending 2.

Something I forgot: if you have two identical files with relative imports, then the these will point to different files. They wouldn't if you deduplicate by file contents.

**Comment by danieltroger** [comment]
> Using parcel@nightly instead of parcel-bundler.

Hey man I updated to parcel 2 some time ago and sometimes I just look at the resulting code targeted for modern browsers and enjoy its beauty. So great work.

But this duplication thing still bothers me. Tbh I don't care if the modules are the same or not because I updated my code, but right now my problem is this:

I have a lib folder with a bunch of code I include everywhere, so we have like
```
~/main_folder $ ls lib
much_code1.ts much_code2.ts much_code3.ts
```

But then I have a module which I made and put in my package.json of my project. So my project is ~/main_folder/project and if I include something from lib, parcel includes it from ~/main_folder/project/node_modules/lib/much_code1.ts

Now the problem is, if I also include the same file in my other module, the path will be this: ~/main_folder/project/node_modules/module_i_made/lib/much_code1.ts

And well my lib folder contains much code, so like 5 files we be included twice and bring up the bundle size. And that bothers me.

Because the "lib" in either of the two paths is a symlink which links to ~/main_folder/lib, IT IS LITERALLY THE SAME FILE. Parcel is just to stupid to notice it because it's a symlink.

It seems like it doesn't include files twice if it doesn't go through a symlink, so please: just resolve the symlinks so I can be happy.

**Comment by mischnic** [comment]
> Now the problem is, if I also include the same file in my other module, the path will be this: ~/main_folder/project/node_modules/module_i_made/lib/much_code1.ts

We do take the realpath in Parcel 2. Or at least will do that once https://github.com/parcel-bundler/parcel/pull/5508 is merged

**Comment by danieltroger** [comment]
Everyone: check if your problems still persists, mine are gone now that #5508 is merged

**Comment by mischnic** [comment]
If someone runs into this agian with Parcel 2: please open a new issue with a code sample.

**Comment by Offirmo** [comment]
Hi @mischnic when is Parcel 2 due to be released? Could we have a blog post from time to time?

**Comment by danieltroger** [comment]
@Offirmo If your only assets are js or ts I'd just recommend to start using it already. AFAIK the only things that aren't complete are some extra features which few people need.

If you want an example of a package.json of v2 you can look at the "reproduction repo" I posted here https://github.com/parcel-bundler/parcel/issues/5555

**Comment by compmaster** [comment]
Confirmed in HTML5 Boilerplate v8.0.0. Bundled index.html contains duplicated .js files with the same content:

```html
<script src="/normalize.1114379f.js"></script><script src="/main.c52e0fe2.js"></script></head>
<body>
<script src="/modernizr-3.11.2.min.659b9f56.js"></script>
<script src="/plugins.7f97feee.js"></script>
<script src="/main.fb6bbcaf.js"></script>
```
Originally in index.html these scripts are included once within `body` tag
```html
<script src="js/vendor/modernizr-3.11.2.min.js"></script>
<script src="js/plugins.js"></script>
<script src="js/main.js"></script>
```

**Comment by nbkhope** [comment]
I'm using `"parcel": "^2.0.1",` and one of my dependencies is symlinked:

```
"dependencies": {
  "@nbkhope/something": "file:../something",
```

And has `@babel/core` installed:

```
"devDependencies": {
   "@babel/core": "^7.16.0",
```

The `package.json` for `@nbkhope/something` also contains `@babel/core`

```
"devDependencies": {
   "@babel/core": "^7.16.0",
```

When I run

```
npx parcel build index.js
```

I get the following error:

```
🚨 Build failed.

@parcel/package-manager: Could not find module "@babel/core", but it was listed in
package.json. Run your package manager first.

  /some/where/something/package.json:18:5
    17 |   "devDependencies": {
  > 18 |     "@babel/core": "^7.16.0",
  >    |     ^^^^^^^^^^^^^ Defined here, but not installed
```

**How can I make parcel look at the node_modules in the root of my project instead of the node_modules of the directory of the package that is imported that is symlinked as `<projectRoot>/node_modules/@nbkhope/something --> ../../../something` ?**

**Comment by MartinDawson** [comment]
> I'm using `"parcel": "^2.0.1",` and one of my dependencies is symlinked:
> 
> ```
> "dependencies": {
>   "@nbkhope/something": "file:../something",
> ```
> 
> And has `@babel/core` installed:
> 
> ```
> "devDependencies": {
>    "@babel/core": "^7.16.0",
> ```
> 
> The `package.json` for `@nbkhope/something` also contains `@babel/core`
> 
> ```
> "devDependencies": {
>    "@babel/core": "^7.16.0",
> ```
> 
> When I run
> 
> ```
> npx parcel build index.js
> ```
> 
> I get the following error:
> 
> ```
> 🚨 Build failed.
> 
> @parcel/package-manager: Could not find module "@babel/core", but it was listed in
> package.json. Run your package manager first.
> 
>   /some/where/something/package.json:18:5
>     17 |   "devDependencies": {
>   > 18 |     "@babel/core": "^7.16.0",
>   >    |     ^^^^^^^^^^^^^ Defined here, but not installed
> ```
> 
> **How can I make parcel look at the node_modules in the root of my project instead of the node_modules of the directory of the package that is imported that is symlinked as `<projectRoot>/node_modules/@nbkhope/something --> ../../../something` ?**

You don't need to do this if you use lerna: ` "@nbkhope/something": "file:../something",`

You should specify it like a normal package. Lerna will then handle the symlinking and this error should not happen.
